pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					repo, ok := c.mirrors.Get(mirrorName)
+
+					hrs, err := c.helmReleasesForMirror(mirrorName)
+					if err != nil {
+						c.logger.Log("error", ErrReleasesForMirror.Error(), "mirror", mirrorName, "err", err)
+						continue
+					}
+
+					// We received a signal from a no longer existing
+					// mirror.
+					if !ok {
+						if len(hrs) == 0 {
+							// If there are no references to it either,
+							// just continue with the next mirror...
+							continue
+						}
+
+						c.logger.Log("warning", ErrNoMirror.Error(), "mirror", mirrorName)
+						for _, hr := range hrs {
+							c.maybeMirror(mirrorName, hr.Spec.GitChartSource.GitURL)
+						}
+						// Wait for the signal from the newly requested mirror...
+						continue
+					}
+
+					c.processChangedMirror(mirrorName, repo, hrs)
+				}
+			case <-stopCh:
+				c.logger.Log("info", "stopping sync of git chart sources")
+				return
+			}
+		}
+	}()
+}
+
+// ChartDir returns a newly exported `GitChartDir` for the given `v1.HelmRelease`
+// at the latest HEAD, or an error.
+func (c *GitChartSync) ChartDir(hr *v1.HelmRelease) (GitChartDir, error) {
+	mirror := mirrorName(hr)
+	repo, ok := c.mirrors.Get(mirror)
+	if !ok {
+		// We did not find a mirror; request one, return, and wait for
+		// signal.
+		c.maybeMirror(mirror, hr.Spec.GitURL)
+		return GitChartDir{}, ChartNotReadyError{ErrNoMirror}
+	}
+
+	s, ok, err := c.sync(hr, mirror, repo)
+	if err != nil {
+		return GitChartDir{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	defer cancel()
+	export, err := repo.Export(ctx, s.head)
+	if err != nil {
+		return GitChartDir{}, ChartUnavailableError{err}
+	}
+
+	return GitChartDir{export, s.head}, nil
+}
+
+// Delete cleans up the source reference for the given `v1.HelmRelease`,
+// this includes the mirror if there is no reference to it from sources.
+// It returns a boolean indicating a successful removal (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) Delete(hr *v1.HelmRelease) bool {
+	c.sourcesMu.Lock()
+	defer c.sourcesMu.Unlock()
+
+	// Attempt to get the source from store.
+	source, ok := c.sources[hr.ResourceID().String()]
+	if ok {
+		// Remove the in store source.
+		delete(c.sources, hr.ResourceID().String())
+
+		if hrs, err := c.helmReleasesForMirror(source.mirror); err == nil && len(hrs) == 0 {
+			// The mirror is no longer in use by any source;
+			// stop and delete the mirror.
+			c.mirrors.StopOne(source.mirror)
+		}
+	}
+	return ok
+}
+
+// SyncMirrors instructs all git mirrors to sync from their respective
+// upstreams.
+func (c *GitChartSync) SyncMirrors() {
```

Comment:
This seems to be needed because of https://github.com/fluxcd/helm-operator/commit/c20e5c84c2e4580c3ab5eba157250c012a1ae9c3 ,  but I cannot find a matching PR. Would you mind clarifying why the http endpoint was added>

Reply:
It is the operators equivalent of `fluxctl sync`, but via an HTTP endpoint, to give users the option to instantly trigger the refresh from git mirrors from e.g. CI, instead of waiting for the poll interval to kick in.

Reply:
Original PR: https://github.com/fluxcd/flux/pull/1776

Reply:
Of course, I forgot this repo was split out recently, sorry.